Use automake

mike-burns opened the pull request:
This introduces a `Makefile.am`, replacing `Makefile.in`. We get a few
benefits from this:

- `dist` and `distcheck` targets.
- Portability, especially to non-GNU systems.
- Simplification.

Automake's policy is that of explicitness; thus, the `Makefile.am` is
long due to the `EXTRA_DIST` included for the tests. This comes into
play in the `distcheck` target, which assumes the tests are distributed
with the program so that package builders can ensure that the system
works locally.

The `check` target makes use of the existing test suite and
infrastructure; little has changed from `Makefile.in`.

While here, remove generated files from our commits.

Sheepishly I must admit: I did not run the full test suite, as it requires find(1) with GNU extensions.

masatake commented:
Thank you for working on this area.

Where is Makefile.am?

mike-burns commented:
The `Makefile.am` is so large that GitHub refuses to show it. This is because of the `EXTRA_DIST`.

https://github.com/universal-ctags/ctags/pull/676/files#diff-c949f93d03f44a4217d7a138f9e2e54a
https://github.com/mike-burns/ctags/blob/automake/Makefile.am

masatake commented:
Let's utilize "include" directive and split Makefile.am smaller parts.
e.g. 

```
include $(top_srcdir)/source.mak
include $(top_srcdir)/testing.mak
...
```
source.mak exits already. Building scripts for othre platforms depend on source.mak.
testing.mak doesn't exist. It should contains targets for testing like units, tmain, and so on.

Temporary, you can ignore files under data and libexec.
Could you focus on
1. making smaller Makefile.am, and
2. building ctags binary on travis  

After achieving the above two goals, I will work on test cases related area with you.
We may want tmain.mak and units.mak, in which test cases are listed. Scripts may be needed.

Thank you.

mike-burns commented:
Thank you for the great feedback.

I have pulled in `sources.mak`. This file had some variable names that conflict with automake variable names, so I renamed them (the HEADER ones). I also removed lines that  did not seem to be used.

I moved all of the `check` target into `testing.mak`.

I introduced `--enable-travis` and `--enable-valgrind` options to the `configure` script. This is because there is no portable way of checking whether an environment variable is defined in a Makefile. This causes the tests to pass again on all but Mingw32. I'll have to look at the Mingw32 failure later: https://travis-ci.org/universal-ctags/ctags/jobs/91216334

mike-burns commented:
I got the builds to pass by:

- Discovering and fixing `mk_mingw.mak` and `mk_mvc.mak`.
- Including `general.h` everywhere that a `boolean` is used.
- Including `stdio.h` everywhere that a `FILE` is used.

ffes commented:
Being a Windows developer I don't really know much about automake and alike. So I will leave reviewing those changes to the others who are more knowledgeable on that subject.

But what does this mean for the commands that need to be used to build ctags? I don't see any changes to `docs/autotools.rst`. Is that because there is no need, or is that because you forgot to update that file?
Since we are hoping to get our first release out any time soon, documentation become more important!

masatake commented:
@mike-burns, I will do cherry-pick your acceptable changes. You may have to do "git push --force" when our master branch is updated with the cherry-picking. I will focus on this issue but I don't have much time this week.

masatake commented on autogen.sh:
```diff
@@ -1,4 +1,7 @@
 echo "Example: "
 echo "	$ autoreconf -f -i -v"
+
+exit 64
```
What is this 64?

masatake replied:
This will be the place to generate testing-cases.mak.

mike-burns replied:
From sysexits(3):

```
 EX_USAGE (64)	   The command was used incorrectly, e.g., with the
			   wrong number of arguments, a bad flag, a bad syntax
			   in a parameter, or whatever.
```

It's the erroneous return value for when instructions are shown.

I ran into issues with autogen.sh returning true (as it did before), because I instinctively ran it, && ./configure && make, and my shell kept running commands as if it had succeeded.

masatake commented:
Trais is passed! Great.

mike-burns commented:
@ffes the documentation is still correct. As a summary, the fundamental change here is: 

Previously: `./configure` used `Makefile.in` to generate `Makefile`.
In this PR: `./configure` uses `Makefile.am` to generate `Makefile.in`, and then uses that to generate `Makefile`.

This is all transparent to the user, mostly thanks to the fact that autoconf and automake are built by the same team.

mike-burns commented:
I split out a `test-cases.mak`, generated from `autogen.sh` (via a helper script in `misc/`).

I briefly went down the path of `testing.mak.in` and having `configure` generate the full file, but this lead to complications around `autoreconf`.

masatake commented:
```
+ bc4955f60337ac6d5cf58eafcebf5220b4b5a8ef Include general.h in kind.h
+ dfc5f980a2a18685fdc9fc72e260c99e903aade6 FILE comes from stdio.h
```

Above two are cherry picked.

masatake commented:
Oh, I just remember that the target name `ctags` is already used in automake.
Is this o.k.?

Quoted from info document of automake:
```   Automake also generates a ‘ctags’ rule that can be used to build
‘vi’-style ‘tags’ files.  The variable ‘CTAGS’ is the name of the
program to invoke (by default ‘ctags’); ‘CTAGSFLAGS’ can be used by the
user to pass additional flags, and ‘AM_CTAGSFLAGS’ can be used by the
‘Makefile.am’.

```

What I got when running autogen.sh.
```
[yamato@x201]~/var/ctags-github% ./autogen.sh 
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force 
autoreconf: configure.ac: tracing
autoreconf: configure.ac: not using Libtool
autoreconf: running: /usr/bin/autoconf --force
autoreconf: running: /usr/bin/autoheader --force
autoreconf: running: automake --add-missing --copy --force-missing
configure.ac:332: installing './compile'
configure.ac:22: installing './config.guess'
configure.ac:22: installing './config.sub'
configure.ac:11: installing './missing'
Makefile.am: installing './depcomp'
/usr/share/automake-1.15/am/tags.am: warning: redefinition of 'ctags' ...
/usr/share/automake-1.15/am/program.am: ... 'ctags$(EXEEXT)' previously defined here
Makefile.am:6:   while processing program 'ctags'
autoreconf: Leaving directory `.'
```

masatake commented:
I tested. Though I got warning, ctags executable was built successfully.

The greatest news is that tar.gz file was successfully created with `make dist`, and ctags executable is successfully built from the .tar.gz file. Surprisingly "make check"  was done successfully, too.

masatake commented on misc/dist-test-cases:
```diff
@@ -0,0 +1,5 @@
+
+echo "EXTRA_DIST = misc/units \\" > "$1"
+find Tmain Units -type f | sed -e 's/$/\\/' >> "$1"
+echo '$(NULL)' >> "$1"
```
I can improve the arguments for find.

b4n replied:
```shell
#!/bin/sh

( echo "EXTRA_DIST = misc/units"
  find Tmain Units -type f ) | sed -e '$!s/$/ \\/' > "$1"
```

b4n replied:
and the redirection should probably rather be done by the caller of the script

b4n replied:
BTW, finding all files is a bad idea, if e.g. there are leftover *.tmp* files in the test trees.

masatake replied:
This is about autogen.sh. So I will use git list-files for listing files under Units and Tmain.

b4n replied:
Hum, that might show another issue: how to add a test case?  and even more so, how to test it before it gets committed?

Though, the question "how to add a test case" is relevant in all cases if we need to list is explicitly somewhere.  It can be as simple as running a script to update something, but it would have to be documented.

masatake replied:
Files gathered through git-list-files are stored to EXTRA_DIST.
Our own "make check" doesn't refer the value of EXTRA_DIST.
So we can run "make check" without updateing EXTRA_DIST.
The way for adding a new test case is not changed.

masatake commented:
Could you rebuild the branch? Some of commits can be squashed now.
I'll make PRs to your branch.

mike-burns commented:
- The `ctags` target does complain but I think it's fine. Next time I get fee time (one night this week) I'd like to see if I can silence that warning.
- Instead of `dist` I recommend `distcheck` -- it's like `dist`, but also runs the tests and makes sure the executable can build from it.
- I have rebased against `master` and force-pushed. Tonight I'll squash and re-word the commits on this branch.

Thank you for your patience and help with this.

ffes commented:
Guys, when going through the `old-docs` for #674, I found `old-docs/INSTALL`. It contains quite some information about how to use `./configure`, including a list of all the flags. Do you want to add the things still useful and relevant to `docs/autotools.rst` in the same process or do we do that in a separate PR?

mike-burns commented:
@ffes I have updated the old `INSTALL` documentation -- it wasn't very out of date afterall. However, part of the appeal of autotools is their popularity -- people who install from source tend to know that `./configure --help` will show all the possible flags. I'll leave it up to the universal-ctags team to decide what should be moved into `docs/`.

Is this ready for me to squash?

k-takata commented on source.mak:
Adding `HEADERS_*` into `SOURCES` causes compilation error with MSVC:
https://ci.appveyor.com/project/masatake/ctags/build/1.0.328/job/b9ekotssx5ywctl7
https://ci.appveyor.com/project/masatake/ctags/build/1.0.328/job/bp2v68a5ja9c69oa

This should be:
`SOURCES = $(MAIN_SRCS) $(PARSER_SRCS) $(DEBUG_SRCS)`

k-takata commented:
msys2 build and cygwin build fail on AppVeyor:
https://ci.appveyor.com/project/masatake/ctags/build/1.0.328/job/g7csj27k6un1kn30
https://ci.appveyor.com/project/masatake/ctags/build/1.0.328/job/oemtnase8o1v4ov5

`win32/appveyor.bat` needs to be updated.

Maybe:

```diff
--- a/win32/appveyor.bat
+++ b/win32/appveyor.bat
@@ -109,7 +109,7 @@ set CHERE_INVOKING=yes
 rem bash -lc "for i in {1..3}; do update-core && break || sleep 15; done"
 rem bash -lc "for i in {1..3}; do pacman --noconfirm -Su mingw-w64-%MSYS2_ARCH%-{gcc,libiconv} automake autoconf make dos2unix && break || sleep 15; done"
 
-bash -lc "autoreconf -vfi"
+bash -lc "./autogen.sh"
 :: Patching configure.
 :: Workaround for "./configure: line 557: 0: Bad file descriptor"
 perl -i".bak" -pe "s/^test -n \".DJDIR\"/#$&/" configure
@@ -139,7 +139,7 @@ c:\cygwin\setup-x86.exe -qnNdO -R C:/cygwin -s http://cygwin.mirror.constant.com
 PATH c:\cygwin\bin;%PATH%
 set CHERE_INVOKING=yes
 bash -lc ""
-bash -lc "autoreconf -vfi"
+bash -lc "./autogen.sh"
 :: Patching configure.
 :: Workaround for "./configure: line 557: 0: Bad file descriptor"
 perl -i".bak" -pe "s/^test -n \".DJDIR\"/#$&/" configure
```

k-takata commented on source.mak:
Ah, now I understand that automake requires all header files to be listed in `SOURCES`.
How about this?

```
ALL_SRCS = $(MAIN_SRCS) $(PARSER_SRCS) $(DEBUG_SRCS)

SOURCES = $(ALL_SRCS) $(HEADERS_ALL)

    ...

OBJECTS = \
	$(ALL_SRCS:.c=.$(OBJEXT)) \
	$(LIBOBJS)
```

`mk_mvc.mak` and `mk_mingw.mak` should be updated to use `ALL_SRCS` instead of `SOURCES`.

k-takata commented on Makefile.in:
```diff
@@ -1,535 +0,0 @@
-
-commit-stamp: FORCE
-	echo "$(REV)" > $@.tmp
-	if diff -q $@ $@.tmp; then rm -f $@.tmp; else mv -f $@.tmp $@; fi
```
Is `commit-stamp` target not available in `Makefile.am`?
I don't see the commit-id in output of `ctags --version`.

masatake replied:
This one will be the final issue in this build-system conversion.
I will work on.

mike-burns commented:
I made use of the `!=` variable assignment operator for the `CTAGS_COMMIT_ID` because `!=` is available in multiple `make` implementations. However, it is only available in GNU Make 4.0+, which is from 2013-10-09. The OS that you've configured Travis CI to use is much older and uses GNU Make 3.81.

Ubuntu upgraded their GNU Make in April 2015 with 15.04/Vivid Vervet. The latest stable is 15.10/Wily Werewolf, from October 2015. Is it reasonable to update Travis CI to use those versions?

mike-burns commented:
@k-takata thank you very much for your help with Win32. I'm curious about the remaining failure:

```
Environment: compiler=msys2, MSYS2_ARCH=x86_64, MSYS2_DIR=msys64, MSYSTEM=MINGW64
```

That seems to be running a normal `./autogen.sh && ./configure && make` instead of pulling in `mk_mvc.mak` or `mk_mingw.mak`. Is that correct? A normal build like that should work under msys2/mingw64? Should I add a `configure.ac` check for `regcomp` and pull in `gnu_regex` if needed?

k-takata commented:
Yes, msys2/mingw64 uses normal autotools.
It seems that msys2/mingw64 doesn't have fnmatch and gnu_regex. So, built-in fnmatch and gnu_regex should be used.

> Should I add a configure.ac check for regcomp and pull in gnu_regex if needed?

gnu_regex and fnmatch are already checked in configure.ac. Is the result not used in Makefile.am?

mike-burns commented:
Thanks @k-takata . Windows tests are passing again.

@masatake what do you think of running the tests with a modern GNU Make, as I described above?

masatake commented:
What we have to do after merging automake releated commits:

* reimplement the way to turn on valgrind at runtime, not at building time,
* reimplement the way to embed the git revision number into binary,
* remove files in data/optlib and introduce my translator, and
* improve the way to enumerate the files under Tmain and Units with git output.

autogen.sh will become important to solve above things.

masatake commented:
Can I take over this?

BTW,
I requested using HEADERS instead of HEADS. I recognized my request is wrong. HEADS is better because it is consistent with SRCS.

mike-burns commented:
@masatake please do. My time to spend on this is sporadic. I'll keep watching this thread in case I can help.

masatake commented:
@mike-burns, thank you.

masatake commented:
@mike-burns, in sources.mak you replaced $(MAIN_DIR) with "main". I wonder why. "DIR" is reserved word in automake?

mike-burns commented:
The `dir` suffix is meaningful to autotools and I wanted to avoid confusion and mistakes. `MAIN_DIR` might be fine, though, due to the underscore.

masatake commented:
@mike-burns, thank you. I will use MAIN_DIR to make the changes smaller.

masatake commented:
@mike-burns, one more question.
In cbc10b5 "Clean up old-docs/INSTALL … " you wrote:
```
    There is no quick and simple way to rename these programs, but there is
    a simple way to add a suffix to the end. Document that and remove the
    other documentation.
```

However, configure --help says: 
```
Program names:
  --program-prefix=PREFIX            prepend PREFIX to installed program names
  --program-suffix=SUFFIX            append SUFFIX to installed program names
  --program-transform-name=PROGRAM   run sed PROGRAM on installed program names
```

It looks for me there are quick and simple ways to rename these programs.
Am I wrong?

mike-burns commented:
You are right! They're not as simple as the prior method, but they are worth documenting.

I wasn't sure why the original documentation mentioned the ability to rename a program. Are people renaming it?

masatake commented:
@mike-burns, thank you. People might want to rename exuberant ctags because there was older classical ctags. See https://en.wikipedia.org/wiki/Ctags.

b4n commented on Makefile.am:
```diff
@@ -0,0 +1,83 @@
+AUTOMAKE_OPTIONS = foreign
```
I'd rather see this in the `AM_INIT_AUTOMAKE` call

masatake replied:
Done in my local tree. Thank you.

b4n commented:
> Oh, I just remember that the target name ctags is already used in automake.
Is this o.k.?

[…]

> The ctags target does complain but I think it's fine. Next time I get fee time (one night this week) I'd like to see if I can silence that warning.

In addition to the warning, this makes Make believe the `ctags` target is never up-to-date, so it'll always keep linking it on each run no matter what:
```console
$ make V=0
make  all-am
make[1]: Entering directory '[…]'
  CCLD     ctags
make[1]: Leaving directory '[…]'
```

It's probably not a large issue, but that's where I stopped when I [attempted the same thing](/b4n/fishman-ctags/commit/e83985eaf24b648ae2c955603d2435cc3e95386a).

b4n commented on configure.ac:
```diff
@@ -526,13 +518,17 @@ PRETTY_ARG_VAR([WARNING_CFLAGS], [C compiler warning flags],
 
+AM_CONDITIONAL([HAVE_FNMATCH], [test "x$have_fnmatch" = "xyes"])
+if test "xno" = "x$have_fnmatch"; then
+    CPPFLAGS="$CPPFLAGS -I${srcdir}/fnmatch"
```
this should be done in the Makefile, and not touch the user variable `CPPFLAGS` (as it is, if the user sets `CPPFLAGS` when invoking Make (which is totally valid and expected to be supported by Autotools' standards), it'll fail miserably if it requires that extra inclusion but the user didn't know)

masatake replied:
My understanding is that we should use per library flag. I will do so.

masatake replied:
Many **_CPPFLAGS" in Makefile.in are gone in Makefile.am.
I have to revive them.

b4n commented on configure.ac:
```diff
@@ -618,12 +614,15 @@ AC_MSG_RESULT($regcomp_works)
 dnl Using bundled regex implementation
 AC_LIBOBJ([regex])
-CPPFLAGS="$CPPFLAGS -I${srcdir}/gnu_regex"
+CPPFLAGS="$CPPFLAGS -I${srcdir}/gnu_regex -D__USE_GNU"
```
this should be fixed at some point too, but well, at least it was already there

b4n commented:
BTW, you should untrack and ignore *install-sh* too.

masatake commented:
> BTW, you should untrack and ignore install-sh too.

Done in my local tree.

mike-burns commented:
Closing in favor of #700.
